data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroInputCodec.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.dataprepper.plugins.codec.avro;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+
+import org.apache.avro.file.DataFileStream;
+import org.apache.avro.util.Utf8;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPlugin;
+import org.opensearch.dataprepper.model.codec.InputCodec;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.log.JacksonLog;
+import org.opensearch.dataprepper.model.record.Record;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+
+
+/**
+ * An implementation of {@link InputCodec} which parses avro records into fields.
+ */
+@DataPrepperPlugin(name="avro", pluginType = InputCodec.class)
+public class AvroInputCodec implements InputCodec {
+
+    private static final Logger LOG =  LoggerFactory.getLogger(AvroInputCodec.class);
+
+    @Override
+    public void parse(InputStream inputStream, Consumer<Record<Event>> eventConsumer) throws IOException {
+
+
+        Objects.requireNonNull(inputStream);
+        Objects.requireNonNull(eventConsumer);
+        parseAvroStream(inputStream, eventConsumer);
+
+    }
+
+    private void parseAvroStream(final InputStream inputStream, final Consumer<Record<Event>> eventConsumer) {
+
+        try {
+
+            final byte[] avroData=inputStream.readAllBytes();
+            ByteArrayInputStream byteArrayInputStream=new ByteArrayInputStream(avroData);
+            DataFileStream<GenericRecord> stream = new DataFileStream<GenericRecord>(byteArrayInputStream, new GenericDatumReader<GenericRecord>());
+            Schema schema=stream.getSchema();
+
+            while (stream.hasNext()) {
+
+                final Map<String, Object> eventData = new HashMap<>();
+                GenericRecord avroRecord= stream.next();
+
+                for(Schema.Field field : schema.getFields()) {
+                    Object value=decodeValueIfEncoded(avroRecord.get(field.name()));
```

Comment:
I think there might be a few other scenarios that need to be addressed here.

The documentation states the following:

> Schema records are implemented as [GenericRecord](https://avro.apache.org/docs/current/api/java/org/apache/avro/generic/GenericRecord.html).
> Schema enums are implemented as [GenericEnumSymbol](https://avro.apache.org/docs/current/api/java/org/apache/avro/generic/GenericEnumSymbol.html).
> Schema arrays are implemented as [Collection](https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html?is-external=true).
> Schema maps are implemented as [Map](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html?is-external=true).
> Schema fixed are implemented as [GenericFixed](https://avro.apache.org/docs/current/api/java/org/apache/avro/generic/GenericFixed.html).
> Schema strings are implemented as [CharSequence](https://docs.oracle.com/javase/8/docs/api/java/lang/CharSequence.html?is-external=true).
> Schema bytes are implemented as [ByteBuffer](https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html?is-external=true).
> Schema ints are implemented as [Integer](https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html?is-external=true).
> Schema longs are implemented as [Long](https://docs.oracle.com/javase/8/docs/api/java/lang/Long.html?is-external=true).
> Schema floats are implemented as [Float](https://docs.oracle.com/javase/8/docs/api/java/lang/Float.html?is-external=true).
> Schema doubles are implemented as [Double](https://docs.oracle.com/javase/8/docs/api/java/lang/Double.html?is-external=true).
> Schema booleans are implemented as [Boolean](https://docs.oracle.com/javase/8/docs/api/java/lang/Boolean.html?is-external=true).
> 

https://avro.apache.org/docs/current/api/java/org/apache/avro/generic/package-summary.html

We want to get everything into common Java types. Which means we should map `GenericRecord`, `GenericEnumSymbol`, and `GenericFixed`.

Here is how I think we can do this:

* `GenericRecord` - This can be solved by recursion. You already have an `avroRecord`. Refactor this to use a recursive method and you should have that one solved.

Something like the following may work. Change this method to look like:

```
while (stream.hasNext()) {
  GenericRecord avroRecord= stream.next();
  final Map<String, Object> eventData = convertRecordToMap(avroRecord, schema);
...
```

Add a new method:

```
Map<String, Object> convertRecordToMap(GenericRecord record, Schema, schema) {
  for(Schema.Field field : schema.getFields()) {
    Object value = avroRecord.get(field.name());
    if(value instanceof GenericRecord) {
      value = convertRecordToMap((GenericRecord)value, schema);
    } else {
       ... Need to handle other conversions, including current UTF-8
    }
}
```

* For `GenericEnumSymbol`, I'm not quite sure the best approach. This might be a scenario where converting to string is appropriate. So, calling `value.toString()` I think will give the right value.

* I'm also not so sure about `GenericFixed`. But, it appears that you need to get the byte array from it.

Please also add unit test cases for each of these scenarios.

Reply:
@umayr-codes ,

As an alternative, in the `decodeValueIfEncoded` method, you can check for unsupported types and throw exceptions for now. We can create a GitHub issue and follow-on with these feature changes in a separate PR.

Either way can work well.

e.g.

```
if(rawValue instanceof GenericEnumSymbol) {
  throw new RuntimeException("The Avro codec does not support enums presently.);
}
```

The current behavior appears it will be passing this through the pipeline, which will cause failures in processors or sinks.

Reply:
This is done for now. Working on the follow up to support all data types.

Reply:
Thank you @umairofficial !